samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java

```diff
@@ -18,47 +18,178 @@
 import org.apache.samza.config.JobCoordinatorConfig;
 import org.apache.samza.config.TaskConfigJava;
+import org.apache.samza.container.SamzaContainer;
 import org.apache.samza.coordinator.JobCoordinator;
 import org.apache.samza.coordinator.JobCoordinatorFactory;
+import org.apache.samza.job.model.ContainerModel;
+import org.apache.samza.job.model.JobModel;
+import org.apache.samza.metrics.JmxServer;
 import org.apache.samza.metrics.MetricsReporter;
 import org.apache.samza.task.AsyncStreamTaskFactory;
 import org.apache.samza.task.StreamTaskFactory;
 import org.apache.samza.util.Util;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
 
 /**
- * StreamProcessor can be embedded in any application or executed in a distributed environment (aka cluster) as an
- * independent process.
  * <p>
- * <b>Usage Example:</b>
- * <pre>
- * StreamProcessor processor = new StreamProcessor(1, config);
- * processor.start();
- * try {
- *  boolean status = processor.awaitStart(TIMEOUT_MS);    // Optional - blocking call
- *  if (!status) {
- *    // Timed out
- *  }
- *  ...
- * } catch (InterruptedException ie) {
- *   ...
- * } finally {
- *   processor.stop();
- * }
- * </pre>
- * Note: A single JVM can create multiple StreamProcessor instances. It is safe to create StreamProcessor instances in
+ *   StreamProcessor can be embedded in any application or executed in a distributed environment (aka cluster) as an
+ *   independent process.
+ * </p>
+ *
+ * <b>Note</b>: A single JVM can create multiple StreamProcessor instances. It is safe to create StreamProcessor instances in
  * multiple threads.
  */
 @InterfaceStability.Evolving
 public class StreamProcessor {
+  private static final Logger LOGGER = LoggerFactory.getLogger(StreamProcessor.class);
+
   private final JobCoordinator jobCoordinator;
-  private final StreamProcessorLifecycleListener lifecycleListener;
-  private final String processorId;
+  private final StreamProcessorLifecycleListener processorListener;
+  private final Object taskFactory;
+  private final Map<String, MetricsReporter> customMetricsReporter;
+  private final Config config;
+  private final long taskShutdownMs;
+  private ExecutorService executorService;
+  private volatile SamzaContainer container = null;
+
+  @VisibleForTesting
+  JobCoordinatorListener jobCoordinatorListener = null;
+
+  // Latch used to synchronize between the JobCoordinator thread and the container thread, when the container is
+  // stopped due to re-balancing
+  private volatile CountDownLatch jcContainerLatch = new CountDownLatch(1);
+  private volatile boolean processorOnStartCalled = false;
+
+  SamzaContainer createSamzaContainer(ContainerModel containerModel, int maxChangelogStreamPartitions, JmxServer jmxServer) {
```

Comment:
I tried it. It'll change a lot of existing files and there are checkstyle errors we'll need to fix. Let's do that after this release to avoid merge conflicts for in-progress PRs.